diff --git a/main.c b/main.c
index 61b8ea7..8aea906 100644
--- a/main.c
+++ b/main.c
@@ -41,7 +41,9 @@
 #include "thumbs.h"
 #include "util.h"
 #include "window.h"
+#define _WINDOW_CONFIG
 #include "config.h"
+#undef _WINDOW_CONFIG
 
 enum {
 	FILENAME_CNT = 1024,
@@ -912,6 +914,26 @@ int main(int argc, char **argv)
 		tns.thumbs = NULL;
 		load_image(fileidx);
 	}
+
+	win.w = WIN_WIDTH;
+	win.h = WIN_HEIGHT;
+
+	if (mode == MODE_IMAGE && filecnt == 1) {
+		win.h -= win.bar.h;
+
+		if (img.w <= win.w && img.h <= win.h) {
+			win.w = img.w;
+			win.h = img.h;
+		} else {
+			double scale = (img.w * win.h > img.h * win.w) ?
+			               (double) win.w / img.w : (double) win.h / img.h;
+			win.w = (int) (img.w * scale);
+			win.h = (int) (img.h * scale);
+		}
+
+		win.h += win.bar.h;
+	}
+
 	win_open(&win);
 	win_set_cursor(&win, CURSOR_WATCH);
 
diff --git a/window.c b/window.c
index ff7f116..7895699 100644
--- a/window.c
+++ b/window.c
@@ -215,12 +215,8 @@ void win_open(win_t *win)
 		                       &win->w, &win->h);
 	if ((gmask & WidthValue) != 0)
 		sizehints.flags |= USSize;
-	else
-		win->w = WIN_WIDTH;
 	if ((gmask & HeightValue) != 0)
 		sizehints.flags |= USSize;
-	else
-		win->h = WIN_HEIGHT;
 	if ((gmask & XValue) != 0) {
 		if ((gmask & XNegative) != 0) {
 			win->x += e->scrw - win->w;
